File: Problem_1.py
"""
Problem : 1

Time Complexity : 
Approach 1 - O(n^2)
Approach 2 - O(nlogn)

Space Complexity : 
Approach 1 - O(n)
Approach 2 - O(n)

Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No


"""

# Longest Increasing Subsequence

# Approach - 1
# DP Array

class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n=len(nums)
        dp=[1 for _ in range(n)]

        for i in range(n):
            for j in range(i):
                if nums[i]>nums[j]:
                    dp[i]=max(dp[i],dp[j]+1)
        return max(dp)

# Approach - 2
# Binary Search
    
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n=len(nums)
        arr=[0 for _ in range(n)]
        arr[0]=nums[0]
        le=1
        for i in range(1,n):
            if nums[i]>arr[le-1]:
                arr[le]=nums[i]
                le+=1
            else:
                bsIdx=self.binarySearch(arr,0,le-1,nums[i])
                arr[bsIdx]=nums[i]
        return le

    
    def binarySearch(self,arr,start,end,num):
        while start<=end:
            mid=start+(end-start)/2
            if arr[mid]==num:
                return mid
            elif arr[mid]>num:
                end=mid-1
            else:
                start=mid+1
        return start

File: Problem_1.js
// ## Problem1: Longest Increasing Subsequence (https://leetcode.com/problems/longest-increasing-subsequence/)

// DP
// Time Complexity : O(n^2)
// Space Complexity : O(n)

// Binary Search
// Time Complexity : O(nlogn)
// Space Complexity : O(n)

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

var binarySearch = (arr, low, high, target) => {
    while (low <= high) {
        let mid = Math.floor(low + ((high - low) / 2));
        if (arr[mid] === target) {
            return mid;
        } else if (arr[mid] > target) {
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
    return low;
}
/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function (nums) {
    if (nums === null || nums.length === 0)
        return 0;
    let n = nums.length;
    let len = 1;
    let arr = new Array(n);
    arr[0] = nums[0];
    for (let i = 1; i < n; i++) {
        if (nums[i] > arr[len - 1]) {
            arr[len] = nums[i];
            len++;
        } else {
            let index = binarySearch(arr, 0, len - 1, nums[i]);
            arr[index] = nums[i];
        }
    }
    return len;

    // let dp = new Array(n);
    // let max = 1;
    // dp.fill(1);
    // for(let i=1; i<n; i++){
    //     for(let j=0; j<i; j++){
    //         if(nums[i]>nums[j]){
    //             dp[i] = Math.max(dp[i], dp[j]+1);
    //             max = Math.max(max, dp[i]);
    //         } 
    //     }
    // }
    // return max;
};